client/allocrunnerv2/taskrunner/vault_hook.go

```diff
@@ -0,0 +1,390 @@
+	cancel context.CancelFunc
+
+	// tokenPath is the path in which to read and write the token
+	tokenPath string
+
+	// alloc is the allocation
+	alloc *structs.Allocation
+
+	// taskName is the name of the task
+	taskName string
+
+	// firstRun stores whether it is the first run for the hook
+	firstRun bool
+
+	// future is used to wait on retrieving a Vault token
+	future *tokenFuture
+}
+
+func newVaultHook(config *vaultHookConfig) *vaultHook {
+	ctx, cancel := context.WithCancel(context.Background())
+	h := &vaultHook{
+		vaultStanza:  config.vaultStanza,
+		client:       config.client,
+		eventEmitter: config.events,
+		lifecycle:    config.lifecycle,
+		updater:      config.updater,
+		alloc:        config.alloc,
+		taskName:     config.task,
+		firstRun:     true,
+		ctx:          ctx,
+		cancel:       cancel,
+		future:       newTokenFuture(),
+	}
+	h.logger = config.logger.Named(h.Name())
+	return h
+}
+
+func (*vaultHook) Name() string {
+	return "vault"
+}
+
+func (h *vaultHook) Prerun(ctx context.Context, req *interfaces.TaskPrerunRequest, resp *interfaces.TaskPrerunResponse) error {
+	// If we have already run prerun before exit early. We do not use the
+	// PrerunDone value because we want to recover the token on restoration.
+	first := h.firstRun
+	h.firstRun = false
+	if !first {
+		return nil
+	}
+
+	// Try to recover a token if it was previously written in the secrets
+	// directory
+	recoveredToken := ""
+	h.tokenPath = filepath.Join(req.TaskDir, allocdir.TaskSecrets, vaultTokenFile)
+	data, err := ioutil.ReadFile(h.tokenPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("failed to recover vault token: %v", err)
+		}
+
+		// Token file doesn't exist
+	} else {
+		// Store the recovered token
+		recoveredToken = string(data)
+	}
+
+	// Launch the token manager
+	go h.run(recoveredToken)
+
+	// Block until we get a token
+	select {
+	case <-h.future.Wait():
+	case <-ctx.Done():
+		return nil
+	}
+
+	h.updater.updatedVaultToken(h.future.Get())
+	return nil
+}
+
+func (h *vaultHook) Poststop(ctx context.Context, req *interfaces.TaskPoststopRequest, resp *interfaces.TaskPoststopResponse) error {
+	// Shutdown any created manager
+	h.cancel()
+	return nil
+}
+
+// run should be called in a go-routine and manages the derivation, renewal and
+// handling of errors with the Vault token. The optional parameter allows
+// setting the initial Vault token. This is useful when the Vault token is
+// recovered off disk.
+func (h *vaultHook) run(token string) {
+	// Helper for stopping token renewal
+	stopRenewal := func() {
+		if err := h.client.StopRenewToken(h.future.Get()); err != nil {
+			h.logger.Warn("failed to stop token renewal", "error", err)
+		}
+	}
+
+	// updatedToken lets us store state between loops. If true, a new token
+	// has been retrieved and we need to apply the Vault change mode
+	var updatedToken bool
+
+OUTER:
+	for {
+		// Check if we should exit
+		select {
+		case <-h.ctx.Done():
```

Comment:
`if h.ctx.Err() != nil { return }` is a more succinct way of checking outside of selects. I've been trying to switch to it.